### DISCUSSION TITLE
We will work with independents or with RUMBus?

### DISCUSSION THREAD

**Discussion by Janiel777** [uprm-inso4116-2024-2025-s1/semester-project-trolley-tracker-app#1] [General]
As you may know, the professor gave us the option to work with the RUMBus group. I would like to discuss whether this is actually a good option or whether this could make it more difficult to get good grades on the milestones. I would prefer to work independently of the RUMBus project so I will say a few things about why. Even so, whoever believes that it is better to work with RUMBus is open to discussion in this thread.

The biggest problem we would have by working independently is that we do not have the way to install dedicated hardware on the trolleys. Since we would not have the financing and most likely we would not be able to do it since the trolleys already have hardware installed by the RUMBus team. Even so, this is not the only way to invent a way to record the location of the trolleys. The other option is to make an application for the drivers to use on their personal phones. This way it is very easy to ask for permissions to access their locations and to be able to save their positions in real time.

To show that this is something very simple, I made a private repository that uses React Native and Expo with the Expo Location libraries to access the device's location and React Native Maps to integrate Google Maps within the application. With no more than 70 lines of code, when running it this is the result on my phone:

https://github.com/user-attachments/assets/71954044-aa2b-4183-aeeb-ed225a05ee9d

The idea of ​​this would be to have a database that saves, in addition to other information, the positions that are sent through the application and then a backend can be made with an endpoint so that the client can access that information. That is an easy way to implement.

Now, if the application grows a lot in terms of users, since each user would make queries to the database using that endpoint that returns the positions, the application could become slow. For that we could implement websokets. Using websokets, drivers send their location to the server. The server updates the database and simultaneously sends the new position to all connected clients via WebSockets. The clients receive the new position and update the map immediately.

In addition, other features could be worked on to make the map look customized, which is possible to implement with the library I mentioned above:
![mapa](https://github.com/user-attachments/assets/5cbfe355-022f-4492-9eec-1b5ec73570bf)

And finally, while we are working on the project and we already have the database and the way to collect the drivers' positions, we would not need the drivers to use our app for us to debug or visualize its operation. It is always an option to create simulated information, which in the end is not real but could be useful to visualize and check the project features.

The discussion is open so that anyone can discuss whether it is better to team up with RUMBus.

**Comment by gabrielpadilla7** [comment]
I believe working independently is also a good idea. That way we can build from the ground up. I think the driver phone tracker idea is great, but we'd need to get permission and even then, if one driver doesn't agree to use the app for tracking, then that might mess up the inner workings of the app. I might be wrong, but it's what I think at the moment, if the drivers don't agree, a simulation is the way to go if there's no other feasible way to do it.

**Comment by AngelMartinez33** [comment]
I think that **we should**, at least, **have some sort of communication with the RUMBus team** to discuss the current state of the app. Here's my reasoning behind this:

We are trying to solve a real-world problem. The problem, which is knowing the current location of the bus, was already discussed in the RUMBus team and there is a current implementation of a software system to tackle the issue. Us choosing to go independently without knowing the current state of the app, their findings, and their overall progress is not that helpful. In a real-world scenario, I think software engineers would prefer to build over what has already been found by others rather than starting again from scratch, which gets us nowhere. We should at least get in contact with RUMBus to know what we can do better to solve this issue.

I propose that we contact them ASAP and communicate that we intend to do this project and that there is the possibility of collaborating. Now knowing that our grades are at stake here, I think the terms of the collaboration should be discussed with the RUMBus team and the professor very carefully to see if it is feasible.

Later, we could decide not to collaborate with them, but at least knowing their progress will allow us to give a helpful, new solution for a real problem.

**Reply by AngelMartinez33** [reply]
In addition, I don't think the drivers will allow access to their phone location. There are a lot of privacy issues involved: 1) how they know when the location is not being shared? 2) Will they have the ability to turn the location off? 3) How are they sure that their location is being managed responsibly?

Also, having that there is existing hardware in the buses, I think is probable that they refuse to install more hardware for the same purposes.

**Reply by MrL9000** [reply]
yea I agree with this sentiment, it's, uhh, analogous to what I just said xd

**Comment by MrL9000** [comment]
I would argue the best way to go forwards *currently* is to hedge our bets by working somewhat independently from the RUMBus team, but keep close contact with them for interview and whatnot, and to make set up the structure of the project for *possible* future compatibility with their services (for example picking codebases that allign with what the RUMBus has already made compatibility for).

**Comment by Keiven-Soto** [comment]
I think the best approach to this project would be to get some guidances from the RUMBUS team and understand how they tackled this project and then integrate their solutions to our project and adapt their ideas in a way that separates us from their work but maintaining a similar concept

**Comment by MarkoSchuetz** [comment]
I think it's clear that you have to communicate with the RUMBus team. Even if you then go your separate way. You might get access to the location data stream... As far as the grades are concerned: you could get a great grade either way, whether you collaborate with the existing project or not. With a perspective on future internship or employment the collaboration would be preferred: working with an existing, long-running project and contributing substantially is much more likely what you'll be doing on a job.

**Comment by Janiel777** [comment]
So, the task remains for the manager to communicate with RUMBus (at least the initial contact) and let us know how we can integrate into their work and what features we can start working on.

**Comment by Yahid1** [comment]
As a manager of the project, I wanted to inform everyone that I am in communication with a Representative of the RUMBus project. I will be meeting with them tomorrow as soon as possible to determine possible features that we maybe be able to implement and assist the team. If for some reason we cannot assist the RUMBus team I will look for information that might be helpful in our development of the Trolley Tracker App.

**Reply by Yahid1** [reply]
In my understanding they are happy to receive help from us. In fact they encourage the behavior as they are searching for people to join their organization.

**Comment by yamayrie** [comment]
There are valid pros and cons to both options. Working with RUMBus is beneficial because we can take advantage of the infrastructure that currently exists, such as the installed trackers, and we can possibly speed up our work by expanding upon the existing project. This approach could save us time and resources, especially considering the permissions are already handled by RUMBus (though I thought this was a hypothetical project and permissions might not be an actual hurdle).

On the other hand, starting from scratch gives us the opportunity to develop the project from the ground up, giving us a variety of tasks and the flexibility to implement our own ideas. If we decide to go independent, one effective approach could be to request access to RUMBus’s location data. This would allow us to maintain a key feature of tracking without the need for dedicated hardware on the trolleys.
